clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java

```diff
@@ -803,149 +805,141 @@ protected boolean isReadyToServe(PartitionConsumptionState partitionConsumptionS
-       * Currently, we make it a sloppy in case Kafka topic have duplicate messages.
-       * TODO: find a better solution
+       * If offset lag threshold is set to -1, time lag threshold will be the only criterion for going online.
        */
-      final long versionTopicPartitionOffset =
-          cachedPubSubMetadataGetter.getOffset(getTopicManager(localKafkaServer), versionTopic, partitionId);
-      isLagAcceptable =
-          versionTopicPartitionOffset <= partitionConsumptionState.getLatestProcessedLocalVersionTopicOffset()
-              + SLOPPY_OFFSET_CATCHUP_THRESHOLD;
-    } else {
-      try {
-        // Looks like none of the short-circuitry fired, so we need to measure lag!
-        long offsetThreshold = hybridStoreConfig.get().getOffsetLagThresholdToGoOnline();
-        long producerTimeLagThresholdInSeconds =
-            hybridStoreConfig.get().getProducerTimestampLagThresholdToGoOnlineInSeconds();
-        String msg = msgForLagMeasurement[partitionId];
-
-        // Log only once a minute per partition.
-        boolean shouldLogLag = !REDUNDANT_LOGGING_FILTER.isRedundantException(msg);
-        /**
-         * If offset lag threshold is set to -1, time lag threshold will be the only criterion for going online.
-         */
-        if (offsetThreshold >= 0) {
-          isLagAcceptable = checkAndLogIfLagIsAcceptableForHybridStore(
-              partitionConsumptionState,
-              measureHybridOffsetLag(partitionConsumptionState, shouldLogLag),
-              offsetThreshold,
-              shouldLogLag,
-              true,
-              0);
-        }
+      if (offsetThreshold >= 0) {
+        isLagAcceptable = checkAndLogIfLagIsAcceptableForHybridStore(
+            partitionConsumptionState,
+            measureHybridOffsetLag(partitionConsumptionState, shouldLogLag),
+            offsetThreshold,
+            shouldLogLag,
+            true,
+            0);
+      }
 
+      /**
+       * If the hybrid producer time lag threshold is positive, check the difference between current time and latest
+       * producer timestamp; ready-to-serve will not be reported until the diff is smaller than the defined time lag threshold.
+       *
+       * If timestamp lag threshold is set to -1, offset lag threshold will be the only criterion for going online.
+       */
+      if (producerTimeLagThresholdInSeconds > 0) {
+        long producerTimeLagThresholdInMS = TimeUnit.SECONDS.toMillis(producerTimeLagThresholdInSeconds);
+        long latestConsumedProducerTimestamp =
+            partitionConsumptionState.getOffsetRecord().getLatestProducerProcessingTimeInMs();
+        if (amplificationFactor != 1) {
+          latestConsumedProducerTimestamp = getLatestConsumedProducerTimestampWithSubPartition(
+              latestConsumedProducerTimestamp,
+              partitionConsumptionState);
+        }
+        boolean timestampLagIsAcceptable = checkAndLogIfLagIsAcceptableForHybridStore(
```

Comment:
It should be checking the time-lag since it gets a time-lag threshold param.
There are some conflicts I will need to rebase, and I think this is to execute time-lag measurement logic instead of offset lag.